src/macros.rs

```diff
@@ -10,11 +10,12 @@ macro_rules! harness {
     ( $( $name:path, $root:expr, $pattern:expr ),+ $(,)* ) => {
         fn main() -> ::std::process::ExitCode {
             let mut requirements = Vec::new();
+            use $crate::autoref::*;
 
             $(
                 requirements.push(
                     $crate::Requirements::new(
-                        $name,
+                        $name.kind().resolve($name),
```

Comment:
I've read the case study and *think* i understand what is happening here but it still feels like black magic :D

Reply:
A little bit yeah! It's basically finding the right `kind()` method based on which trait matches, and then calling `resolve` on it.

Reply:
> It's basically finding the right kind() method based on which trait matches
Specifically that there's only one such `kind()` method for each possible type that could be provided here (hence the various trait impls that don't overlap)